README.md
![ubuntu-focal](https://img.shields.io/badge/ubuntu-20.04-red)
![ubuntu-hirsute](https://img.shields.io/badge/ubuntu-21.04-red)
![ubuntu-jammy](https://img.shields.io/badge/ubuntu-22.04-red)
![hamonikr-hanla](https://img.shields.io/badge/hamonikr-hanla-purple)
![hamonikr-taebaek](https://img.shields.io/badge/hamonikr-taeback-green)

# hamonikr-grub-theme

 * HamoniKR-ME (>= 1.4) 지원
 * 사용자 해상도 자동 감지 후 테마 적용
 * 와이드 스크린의 경우 자동으로 비율을 계산
 * Used `NanumSquare Regular 16`

## Install

### 하모니카 OS (>=5.0)
```
wget -qO- https://update.hamonikr.org/add-update-repo.apt | sudo -E bash -

sudo apt install hamonikr-grub-theme
```

### Other Ubutu base distro. (Ubuntu >=20.04)
Release 페이지에서 배포한 최신 패키지를 다운로드 받아서 아래와 같이 설치.

sudo apt install -f ./hamonikr-grub-theme_*_amd64.deb

### Install from source
다운로드 받은 디렉토리 안으로 이동하여 아래 명령어를 입력

```sudo ./install.sh```

설치 후 시스템을 재시작 하면 grub 테마가 적용됩니다.

 * 테마 화면이 보이지 않는 경우 grub 설정을 확인하세요. (하모니카의 기본값은 hidden)
 ```cat /etc/default/grub | grep GRUB_TIMEOUT_STYLE```
 * 위 명령어의 결과가 hidden 으로 되어 있는 경우 grub 화면이 숨겨져서 보이지 않습니다.
 * ```#GRUB_TIMEOUT_STYLE=hidden``` 처럼 해당 줄의 맨 앞에 '#' 표시를 추가하고 저장 후 ```sudo update-grub``` 으로 grub 이미지를 재생성 후 시스템을 재시작하세요.


### How to create grub font
```
sudo grub-mkfont --output=/boot/grub/fonts/NanumGothicCoding.pf2 --size=32 ~/.local/share/fonts/NanumGothicCoding.ttf
```

### How to convert png from icons-svg folder
Use convert-svg-app (https://github.com/hamonikr/convert-svg)
```
convert-svg-app
```

# License
 * GPL v3
 

 # 이슈 또는 버그
 사용 중 문제를 발견하시면 root@hamonikr.org 또는 https://groups.google.com/forum/m/#!forum/hamonikr 에서 알려주세요.


debian/changelog
hamonikr-grub-theme (3.3.2) stable; urgency=medium

  * Release for HamoniKR 7.0
   - Add Exception handling in postinst

 -- HamoniKR <pkg@hamonikr.org>  Thu, 19 Oct 2023 20:49:57 +0900

hamonikr-grub-theme (3.3.1) stable; urgency=medium

  * Relase for Ubuntu 22.04
   - fixed check logic with hamonikr-system in postinst

 -- HamoniKR <pkg@hamonikr.org>  Fri, 08 Sep 2023 12:35:17 +0900

hamonikr-grub-theme (3.3.0) stable; urgency=medium

  * Relase for Ubuntu 22.04

 -- HamoniKR <pkg@hamonikr.org>  Fri, 08 Sep 2023 12:35:17 +0900

hamonikr-grub-theme (3.2.0) unstalbe; urgency=medium

  * Relase for HamoniKR hx
   - Build on LDME 5 i386

 -- HamoniKR <pkg@hamonikr.org>  Fri, 02 Dec 2022 14:12:03 +0900

hamonikr-grub-theme (3.1.9) taebaek; urgency=medium

  * Support LDME and debian bullseye
   - Update for LDME 5

 -- HamoniKR <pkg@hamonikr.org>  Fri, 18 Nov 2022 00:42:46 +0900

hamonikr-grub-theme (3.1.8) taebaek; urgency=medium

  * Changed new background
   - HamoniKR 6.0 LTS taebaek version

 -- HamoniKR <pkg@hamonikr.org>  Thu, 27 Oct 2022 00:49:09 +0900

hamonikr-grub-theme (3.1.7) taebaek; urgency=medium

  * Add exception handling in postinst

 -- HamoniKR <pkg@hamonikr.org>  Wed, 26 Oct 2022 14:03:29 +0900

hamonikr-grub-theme (3.1.6) taebaek; urgency=medium

  * Changed font

 -- HamoniKR <pkg@hamonikr.org>  Tue, 25 Oct 2022 17:48:27 +0900

hamonikr-grub-theme (3.1.5) taebaek; urgency=medium

  * Release new grub theme
   - include Korean env

 -- HamoniKR <pkg@hamonikr.org>  Mon, 24 Oct 2022 14:24:43 +0900

hamonikr-grub-theme (3.1.4) taebaek; urgency=medium

  * Change background image setting
   - desktop-image-scale-method: "stretch"

 -- HamoniKR <pkg@hamonikr.org>  Mon, 24 Oct 2022 10:59:08 +0900

hamonikr-grub-theme (3.1.3) taebaek; urgency=medium

  * Change background image as black
   - high resolution image (2560x1600)

 -- HamoniKR <pkg@hamonikr.org>  Sun, 23 Oct 2022 22:30:27 +0900

hamonikr-grub-theme (3.1.2) taebaek; urgency=medium

  * Change background image
  * Add propertie : desktop-image-scale-method: "crop"

 -- HamoniKR <pkg@hamonikr.org>  Sun, 23 Oct 2022 15:38:53 +0900

hamonikr-grub-theme (3.1.1) taebaek; urgency=medium

  * Update for HIDPI resolution
  * Forced set resolution if grater than width 1920 px

 -- HamoniKR <pkg@hamonikr.org>  Sat, 15 Oct 2022 20:04:06 +0900

hamonikr-grub-theme (3.1.0) taebaek; urgency=medium

  * Release for HamoniKR 6.0
  * Fixed wrong grub tile variable in debian/postinst
  * Removed hamonikr-info dependancy

 -- HamoniKR <pkg@hamonikr.org>  Sat, 08 Oct 2022 21:23:54 +0900

hamonikr-grub-theme (3.0.2) hanla; urgency=medium

  * Add pre-depends hamonikr-info

 -- HamoniKR <pkg@hamonikr.org>  Tue, 19 Oct 2021 18:56:13 +0900

hamonikr-grub-theme (3.0.2) hanla; urgency=medium

  * Fix prerm

 -- HamoniKR <pkg@hamonikr.org>  Thu, 07 Oct 2021 19:22:16 +0900

hamonikr-grub-theme (3.0.1) hanla; urgency=medium

  * Fix postinst

 -- HamoniKR <pkg@hamonikr.org>  Thu, 09 Sep 2021 16:06:24 +0900

hamonikr-grub-theme (3.0.0) hanla; urgency=medium

  * Release for HamoniKR 5.0

 -- HamoniKR <pkg@hamonikr.org>  Thu, 08 Apr 2021 11:55:44 +0900

hamonikr-grub-theme (2.0.0) stable; urgency=medium

  * Release for hamonikr-server 2.0
   - update theme

 -- HamoniKR <pkg@hamonikr.org>  Fri, 05 Mar 2021 11:40:24 +0900

grub-theme-hamonikr-jin (1.1.0) jin; urgency=medium

  * Updated boot/grub/themes/hamonikr-jin/theme.txt
   - Changed font setting and add new fonts
   - NanumSquare Regular 16

 -- HamoniKR <pkg@hamonikr.org>  Sun, 08 Nov 2020 16:22:21 +0900

grub-theme-hamonikr-jin (1.0.1) jin; urgency=medium

  * Updated boot/grub/themes/hamonikr-jin/theme.txt
   - Changed terminal font

 -- HamoniKR <pkg@hamonikr.org>  Mon, 02 Nov 2020 16:21:24 +0900

grub-theme-hamonikr-jin (1.0.0) jin; urgency=medium

  * Initial release

 -- HamoniKR <pkg@hamonikr.org>  Thu, 29 Oct 2020 17:58:27 +0900


install.sh
#!/bin/bash

# grub-theme-hamonikr
# Copyright (C) 2019 Kevin Kim (root@hamonikr.org)
# - 사용자 해상도 자동 감지 후 테마 적용
# - 와이드 스크린의 경우 자동으로 비율을 계산
# - 한글 폰트로 grub 테마 제작
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

ROOT_UID=0
THEME_DIR="/boot/grub/themes"
THEME_DIR_2="/boot/grub2/themes"
THEME_NAME=hamonikr

# Welcome message
echo -e "\n\t*****************************\n\t*  ${THEME_NAME} - Grub2 Theme  *\n\t*****************************"

# Check command avalibility
function has_command() {
  command -v $1 > /dev/null
}

echo -e "\nChecking for root access..."

# Checking for root access and proceed if it is present
if [ "$UID" -eq "$ROOT_UID" ]; then

  # Create themes directory if not exists
  echo -e "Checking for the existence of themes directory..."
  [[ -d ${THEME_DIR}/${THEME_NAME} ]] && rm -rf ${THEME_DIR}/${THEME_NAME}
  [[ -d ${THEME_DIR_2}/${THEME_NAME} ]] && rm -rf ${THEME_DIR_2}/${THEME_NAME}
  [[ -d /boot/grub ]] && mkdir -p ${THEME_DIR}
  [[ -d /boot/grub2 ]] && mkdir -p ${THEME_DIR_2}

  # Copy theme
  echo -e "Installing ${THEME_NAME} theme..."
  [[ -d /boot/grub ]] && cp -a boot/grub/themes/${THEME_NAME} ${THEME_DIR}
  [[ -d /boot/grub2 ]] && cp -a boot/grub/themes/${THEME_NAME} ${THEME_DIR_2}

  # Set theme
  echo -e "Setting ${THEME_NAME} as default..."
  grep "GRUB_THEME=" /etc/default/grub 2>&1 >/dev/null && sed -i '/GRUB_THEME=/d' /etc/default/grub

  [[ -d /boot/grub ]] && echo "GRUB_THEME=\"${THEME_DIR}/${THEME_NAME}/theme.txt\"" >> /etc/default/grub
  [[ -d /boot/grub2 ]] && echo "GRUB_THEME=\"${THEME_DIR_2}/${THEME_NAME}/theme.txt\"" >> /etc/default/grub

  # 사용중인 해상도를 감지해서 적합한 크기로 조정
  # https://en.wikipedia.org/wiki/Display_resolution
  RATIO=`./get_resolution.sh`
  RESOLUTION="1920x1200"
  [[ "$RATIO" == "16:9" ]] && RESOLUTION="1920x1080"

  # 실제 모니터의 해상도를 적용하려면 아래 처럼 사용
  RESOLUTION=`xrandr --current | grep "*" | head -n 1 | awk -F" " '{print $1}'`

  sed -i.bak "s/.*GRUB_GFXMODE.*/GRUB_GFXMODE=${RESOLUTION}/g" /etc/default/grub
  
  # Update grub config
  echo -e "Updating grub config..."
  if has_command update-grub; then
    update-grub
  elif has_command grub-mkconfig; then
    grub-mkconfig -o /boot/grub/grub.cfg
  elif has_command grub2-mkconfig; then
    grub2-mkconfig -o /boot/efi/EFI/fedora/grub.cfg
  fi

  # Success message
  echo -e "\n\t          ***************\n\t          *  All done!  *\n\t          ***************\n"

else
    # Error message
    echo -e "\n\t   ******************************\n\t   *  Error! -> Run me as root  *\n\t   ******************************\n"
fi


fonts/convert-grub-font.sh
#!/bin/bash
#
# Usage : $0 FONT SIZE

# filename : ${1##*/}
# filename without extension : ${1##*.}  OR ${1%%.*}
# ref : https://linuxhint.com/read_filename_without_extension_bash

[ ! -d ./output ] && mkdir output

if [ "$1" = "all" ] ; then
	for font in $(ls *.ttf *.ttc)
	do
		echo "$font"
		grub-mkfont --output="./output/${font%%.*}-16.pf2" --size=16 "$font"
		file ./output/${font%%.*}-16.pf2
		grub-mkfont --output="./output/${font%%.*}-18.pf2" --size=18 "$font"
		file ./output/${font%%.*}-18.pf2
		grub-mkfont --output="./output/${font%%.*}-22.pf2" --size=22 "$font"
		file ./output/${font%%.*}-22.pf2
		grub-mkfont --output="./output/${font%%.*}-24.pf2" --size=24 "$font"
		file ./output/${font%%.*}-24.pf2
	done

	exit
fi


if [ -z "$1" ] || [ -z "$2" ] ; then
	echo "Require font name and size. ex) $0 ./original.ttf 16" 
	exit
fi

grub-mkfont --output="./output/${1%%.*}-${2}.pf2" --size=$2 "$1"

if [ $? -eq 0 ] ; then
	echo "Output : $( realpath ./output/${1%%.*}-${2}.pf2 )"
	ls -al ./output
else
    echo "Failed."
fi

